backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt

```diff
@@ -75,55 +75,127 @@ internal val IrClass.isArrayWithFixedSizeItems: Boolean
     }
+    ParameterIndex.DISPATCH_RECEIVER_INDEX -> TypeKind.fromType(dispatchReceiverParameter?.type)
+    ParameterIndex.EXTENSION_RECEIVER_INDEX -> TypeKind.fromType(extensionReceiverParameter?.type)
+    else -> TypeKind.fromType(this.valueParameters[index.unmap()].type)
 }
 
-private fun IrFunction.needBridgeToAt(target: IrFunction, index: Int)
-        = hasValueTypeAt(index) xor target.hasValueTypeAt(index)
+private fun IrFunction.hasValueTypeAt(index: ParameterIndex) = typeKindAt(index) is TypeKind.ValueType
+private fun IrFunction.hasReferenceAt(index: ParameterIndex) = typeKindAt(index) is TypeKind.Reference
+
+private fun IrFunction.needBridgeToAt(target: IrFunction, index: ParameterIndex)
+        = bridgeDirectionToAt(target, index).kind != BridgeDirectionKind.NONE
+
+@Suppress("EXPERIMENTAL_FEATURE_WARNING")
+private inline class ParameterIndex(val index: Int) {
+    companion object {
+        val RETURN_INDEX = ParameterIndex(0)
+        val DISPATCH_RECEIVER_INDEX = ParameterIndex(1)
+        val EXTENSION_RECEIVER_INDEX = ParameterIndex(2)
+
+        fun map(index: Int) = ParameterIndex(index + 3)
+
+        fun allParametersCount(irFunction: IrFunction) = irFunction.valueParameters.size + 3
+
+        inline fun forEachIndex(irFunction: IrFunction, block: (ParameterIndex) -> Unit) =
+                (0 until allParametersCount(irFunction)).forEach { block(ParameterIndex(it)) }
+    }
 
-internal fun IrFunction.needBridgeTo(target: IrFunction)
-        = (0..this.valueParameters.size + 2).any { needBridgeToAt(target, it) }
+    fun unmap() = index - 3
+}
 
-internal enum class BridgeDirection {
-    NOT_NEEDED,
-    FROM_VALUE_TYPE,
-    TO_VALUE_TYPE
+internal fun IrFunction.needBridgeTo(target: IrFunction): Boolean {
+    ParameterIndex.forEachIndex(this) {
+        if (needBridgeToAt(target, it)) return true
+    }
+    return false
 }
 
-private fun IrFunction.bridgeDirectionToAt(target: IrFunction, index: Int)
-       = when {
-            hasValueTypeAt(index) && target.hasReferenceAt(index) -> BridgeDirection.FROM_VALUE_TYPE
-            hasReferenceAt(index) && target.hasValueTypeAt(index) -> BridgeDirection.TO_VALUE_TYPE
-            else -> BridgeDirection.NOT_NEEDED
+internal enum class BridgeDirectionKind {
+    NONE,
+    BOX,
+    UNBOX
+}
+
+internal data class BridgeDirection(val irClass: IrClass?, val kind: BridgeDirectionKind) {
+    companion object {
+        val NONE = BridgeDirection(null, BridgeDirectionKind.NONE)
+    }
+}
+
+private fun IrType.isNullableNothing() =
+        isNullable() && classifierOrNull?.isClassWithFqName(KotlinBuiltIns.FQ_NAMES.nothing) == true
+
+private fun IrFunction.bridgeDirectionToAt(overriddenFunction: IrFunction, index: ParameterIndex): BridgeDirection {
+    val irClass = overriddenFunction.typeKindAt(index).irType?.erasure()
+    return when {
+        index == ParameterIndex.RETURN_INDEX && returnType.isNothing() && !overriddenFunction.returnType.isNothing() ->
```

Comment:
Still `isNothing` and other cases checked here. Is this required?

Reply:
These are special cases this entire PR was originally created for. I don't see how to move them to, say, `TypeKind`, because they depend on both types (`this` and `overriddenFunction`). The only alternative would be to add `TypeKind.Nothing`, `TypeKind.NullableNothing` and `TypeKind.Pointer`. What do you think, will it be better? I myself think either one will do.

Reply:
I still don't find handling `Nothing` here necessary.
For example, if you have `Nothing` vs `Any` types, then `TypeKind`s would be different, and the generic machinery would decide that a boxing/unboxing is required. Do you see any issues with this?

Reply:
Ok, this one is true, but what about `Nothing` vs `Int`? They both will become `TypeKind.ValueType`, but a bridge is needed.

Reply:
Here is the table for return slot:
```
return:
    Nothing, Nothing -> none
    Nothing, [-Nothing] -> unbox
    Nothing?, Pointer -> unbox
    Unit, Unit -> none
    Unit, * -> box
    [-Nothing, -Unit], ValueType -> unbox
    ValueType, ValueType -> none
    ValueType, Ref -> box
```
Probably it will look ugly however we implement it.

Reply:
Hmm, now that I am looking at it, looks like the row `Nothing?, Pointer -> unbox` can be removed..

Reply:
Yeah, I think, I've got it.